feedback-examples/streamlit-agent/main.py

```diff
@@ -0,0 +1,169 @@
+from langchain.tools.ddg_search.tool import DuckDuckGoSearchResults
+from langchain.tools.render import format_tool_to_openai_function
+from langsmith import Client
+from streamlit_feedback import streamlit_feedback
+
+st.set_page_config(
+    page_title="Streamlit Agent with LangSmith",
+    page_icon="🦜️️🛠️",
+)
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+client = Client()
+
+st.subheader("🦜🛠️ Ask the bot some questions")
+
+llm = ChatOpenAI(model="gpt-3.5-turbo", temperature=0)
+
+tools = [
+    DuckDuckGoSearchResults(
+        name="duck_duck_go"
+    ),  # General internet search using DuckDuckGo
+]
+
+prompt = ChatPromptTemplate.from_messages(
+    [
+        ("system", "You are a helpful assistant. Current date: {time}"),
+        MessagesPlaceholder(variable_name="chat_history"),
+        ("user", "{input}"),
+        MessagesPlaceholder(variable_name="agent_scratchpad"),
+    ]
+).partial(time=lambda: datetime.utcnow().strftime("%Y-%m-%d %H:%M:%S UTC"))
+
+MEMORY = ConversationBufferMemory(
+    chat_memory=StreamlitChatMessageHistory(key="langchain_messages"),
+    return_messages=True,
+    memory_key="chat_history",
+)
+agent = (
+    {
+        "input": lambda x: x["input"],
+        "agent_scratchpad": lambda x: format_to_openai_functions(
+            x["intermediate_steps"]
+        ),
+        "chat_history": lambda x: x.get("chat_history") or [],
+    }
+    | prompt
+    | llm.bind_functions(functions=[format_tool_to_openai_function(t) for t in tools])
+    | OpenAIFunctionsAgentOutputParser()
+)
+
+agent_executor = AgentExecutor(agent=agent, tools=tools, handle_parsing_errors=True)
+
+
+def _submit_feedback(user_response: dict, emoji=None, run_id=None):
+    score = {"👍": 1, "👎": 0}.get(user_response.get("score"))
+    client.create_feedback(
+        run_id=run_id,
+        key=user_response["type"],
+        score=score,
+        comment=user_response.get("text"),
+        value=user_response.get("score"),
+    )
+    return user_response
+
+
+if st.sidebar.button("Clear message history"):
+    MEMORY.clear()
+
+feedback_kwargs = {
+    "feedback_type": "thumbs",
+    "optional_text_label": "Rate this response in LangSmiths",
+}
+if "feedback_key" not in st.session_state:
+    st.session_state.feedback_key = 0
+
+messages = st.session_state.get("langchain_messages", [])
+for i, msg in enumerate(messages):
+    avatar = "🦜" if msg.type == "ai" else None
+    with st.chat_message(msg.type, avatar=avatar):
+        st.markdown(msg.content)
+    if msg.type == "ai":
+        feedback_key = f"feedback_{int(i/2)}"
+
+        if feedback_key not in st.session_state:
+            st.session_state[feedback_key] = None
+
+        disable_with_score = (
+            st.session_state[feedback_key].get("score")
+            if st.session_state[feedback_key]
+            else None
+        )
+        # This actually commits the feedback
+        streamlit_feedback(
```

Comment:
A lot of the complexity here is in logging feedback with this nicer widget, but we could use a streamlit emoji button for an uglier but simpler implementation I think